### PR TITLE
Clambda layout fixes

### PR DESCRIPTION
**Pull request by mshinwell**
We realised some layout propagation was missing for projection of value slots not of kind Value.

Includes #1202

### CODE REVIEW

**Comment by mshinwell** [comment]
This also increases precision of `machtype`s for field access generated in `Cmmgen`.

File: middle_end/closure/closure.ml [hunk 1]
```diff
@@ -943,7 +948,7 @@ let check_constant_result ulam approx =
           let glb =
             Uprim(P.Pread_symbol id, [], Debuginfo.none)
           in
-          Uprim(P.Pfield i, [glb], Debuginfo.none), approx
+          Uprim(P.Pfield (i, Lambda.layout_any_value), [glb], Debuginfo.none), approx
```
Comment (chambart) [inline middle_end/closure/closure.ml:951]: I'd like to have a `Lambda.layout_symbol` (which would be weird in lambda...) and use that here (and below)

**Review by lthls — APPROVED** [review]

File: middle_end/flambda/extract_projections.ml [hunk 2]
```diff
@@ -125,7 +125,7 @@ let rec analyse_expr ~which_variables expr =
         when Variable.Map.mem move.closure which_variables ->
       projections :=
         Projection.Set.add (Move_within_set_of_closures move) !projections
-    | Prim (Pfield field_index, [var], _dbg)
+    | Prim (Pfield (field_index, _), [var], _dbg)
```
Comment (chambart) [inline middle_end/flambda/extract_projections.ml:128]: This should not ignore the kind: only kind value should be allowed here

Comment (chambart) [reply middle_end/flambda/extract_projections.ml:128]: This might be possible to handle more cases, but it's probably too much work

**Comment by chambart** [comment]
In fact there should be no Pfield with layout not Pvalue if flambda

**Review by chambart — APPROVED** [review]

**Pull request merged**